scene/resources/curve.cpp

```diff
@@ -1496,76 +1499,138 @@ void Curve3D::_bake() const {
+		if (d > CMP_EPSILON) { // avoid the degenerate case of two very close points
+			position = npp;
+			Plane post;
+			post.normal = position;
+			post.d = points[i + 1].tilt;
 
-		dist += d;
+			dist += d;
 
-		pointlist.push_back(post);
-		distlist.push_back(dist);
+			pointlist.push_back(post);
+			distlist.push_back(dist);
+		}
 	}
 
 	baked_max_ofs = dist;
 
-	baked_point_cache.resize(pointlist.size());
-	Vector3 *w = baked_point_cache.ptrw();
-	int idx = 0;
+	const int point_count = pointlist.size();
+	{
+		baked_point_cache.resize(point_count);
+		Vector3 *w = baked_point_cache.ptrw();
 
-	baked_tilt_cache.resize(pointlist.size());
-	real_t *wt = baked_tilt_cache.ptrw();
+		baked_tilt_cache.resize(point_count);
+		real_t *wt = baked_tilt_cache.ptrw();
 
-	baked_up_vector_cache.resize(up_vector_enabled ? pointlist.size() : 0);
-	Vector3 *up_write = baked_up_vector_cache.ptrw();
+		baked_dist_cache.resize(point_count);
+		real_t *wd = baked_dist_cache.ptrw();
 
-	baked_dist_cache.resize(pointlist.size());
-	real_t *wd = baked_dist_cache.ptrw();
+		int idx = 0;
+		for (const Plane &E : pointlist) {
+			w[idx] = E.normal;
+			wt[idx] = E.d;
+			wd[idx] = distlist[idx];
 
-	Vector3 sideways;
-	Vector3 up;
-	Vector3 forward;
+			idx++;
+		}
+	}
+
+	if (!up_vector_enabled) {
+		baked_up_vector_cache.resize(0);
+		return;
+	}
 
-	Vector3 prev_sideways = Vector3(1, 0, 0);
-	Vector3 prev_up = Vector3(0, 1, 0);
-	Vector3 prev_forward = Vector3(0, 0, 1);
+	// Step 2: Caculate up vectors and whole local reference frames
+	//
+	// see C. Dougan, The Parallel Transport Frame, Game Programming Gems 2 for example
+	// for a discussion about why not Frenet frame.
+	{
+		PackedVector3Array baked_forward_vector_cache;
 
-	for (const Plane &E : pointlist) {
-		w[idx] = E.normal;
-		wt[idx] = E.d;
-		wd[idx] = distlist[idx];
+		baked_up_vector_cache.resize(point_count);
+		baked_forward_vector_cache.resize(point_count);
 
-		if (!up_vector_enabled) {
-			idx++;
-			continue;
+		Vector3 *up_write = baked_up_vector_cache.ptrw();
+		Vector3 *forward_write = baked_forward_vector_cache.ptrw();
+
+		const Vector3 *points_ptr = baked_point_cache.ptr();
+		//const real_t *tilts_ptr = baked_tilt_cache.ptr();
+		//const real_t *dists_ptr = baked_dist_cache.ptr();
+
+		Basis frame; // x is right, y is up, z is forward
+		Basis frame_prev;
+
+		// Set initial frame based on y-up rule
+		{
+			Vector3 up(0, 1, 0);
+			Vector3 forward = (points_ptr[1] - points_ptr[0]).normalized();
+			if (forward.is_equal_approx(Vector3())) {
+				forward = Vector3(1, 0, 0);
+			}
+
+			if (abs(forward.dot(up)) > 1.0 - UNIT_EPSILON) {
+				frame_prev = Basis::looking_at(-forward, up);
+			} else {
+				frame_prev = Basis::looking_at(-forward, Vector3(1, 0, 0));
+			}
+
+			up_write[0] = frame_prev.get_column(1);
+			forward_write[0] = frame_prev.get_column(2);
 		}
 
-		forward = idx > 0 ? (w[idx] - w[idx - 1]).normalized() : prev_forward;
+		// Calculate Parallel Transport Frame
```

Comment:
```suggestion
		// Calculate the Parallel Transport Frame.
```

This comment isn't that useful?

Reply:
done